### PR TITLE
API: expose public testing functions in pandas.testing (GH9895)

### LINKED ISSUE #9895: API: add testing functions to public API ?
A lot of other projects that use pandas will (like to) use pandas testing functionality like `assert_frame_equal` in their test suite. Although the pandas testing functions are available in the namespace (#6188), they are not really 'officially' labeled as public API that other projects can use (and rely upon). 
Numpy has a similar submodule  `numpy.testing` (http://docs.scipy.org/doc/numpy/reference/routines.testing.html)

Some things we could do:
- make a selection of the functions in `util.testing` that we want to label as public
- add this list somewhere to the docs
- write docstrings for these public ones (the other could use that as well of course ..)
- add some tests for the public API
- I would also import them into a `pandas.testing` module, so it is this one we can publicly advertise (and users are less tempted to use other non-public functions in the `pandas.util.testing` namespace)

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #9895

This just exposes a few of our testing functions publicly in `pandas.testing`.

- Alternative is to move the full testing module to `pandas.testing` (the actual code), question is then however how to only expose a subset of it publicly.
- I now only included the 3 most obvious ones (frame, series, index). Are there other ones that people would want to use? (the sparse ones?)
- We have quite a lot of keywords to those assert_* functions. We could op to only expose a subset of those publicly (redefining in pandas.testing instead of importing and only passing subset of the kwargs through)

### CODE REVIEW

**Comment by jreback** [comment]
lgtm. need some whatsnew docs (but can prob add onto the ``pandas.errors`` section maybe (rename it though)

**Comment by jorisvandenbossche** [comment]
I did a quick search for which `pandas.util.testing` functionality is used by other projects (apart from the main 3 assert_frame/series/index_equal ones):

- dask:
  - makeDataFrame, makeTimeDataFrame, makeMissingDataframe, makeDateIndex, makeTimedeltaIndex, makeTimeSeries
  - assertRaisesRegexp, assertRaises, assert_produces_warning, assert_numpy_array_equal
- statsmodels:
  - makeDataFrame, makeTimeDataFrame, assert_produces_warning
- zipline:
  - assert_panel_equal (but since panel is deprecated, this is less important)

I am not sure whether we want to commit to keeping some of the `make..` functions stable? (and projects can of course keep using them at own risk as it is now)

For the raises and warning ones, since those are not pandas-objects specific test functions, I would also not include them in the official public ones?

**Comment by jreback** [comment]
xref https://github.com/pandas-dev/pandas/issues/15911

I think we will create a ``pandas.util.benchmarks`` sub-module that holds things like ``make*`` which we can make public, keeping  ``pandas.util.testing`` private (or dask can reach in, doesn't really matter). I don't think these need to be really public anyhow.

cc @mrocklin

**Comment by jreback** [comment]
``makeDataFrame`` in camelCase is a really old naming convention anyhow.

**Comment by jreback** [comment]
pls rebase. lgtm. let's merge and can always amend/add to this.

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -166,6 +166,21 @@ The following are now part of this API:
     'UnsortedIndexError',
     'UnsupportedFunctionCall']
 
+
+.. _whatsnew_0200.enhancements.testing:
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: I think both of these should be moved to the ``privacy`` sub-section now.

Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.20.0.txt]: indeed, will need some general editing

**Comment by Rufflewind** [comment]
Does this mean pytest is now a dependency of pandas?

I just upgraded pandas through [AUR](https://aur.archlinux.org/packages/python-pandas-git/) and got this error message:

```  
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/pandas/__init__.py", line 60, in <module>
    import pandas.testing
  File "/usr/lib/python3.6/site-packages/pandas/testing.py", line 7, in <module>
    from pandas.util.testing import (
  File "/usr/lib/python3.6/site-packages/pandas/util/testing.py", line 22, in <module>
    import pytest
ModuleNotFoundError: No module named 'pytest'
```

I do not have pytest installed.  Is it now required?

**Comment by jreback** [comment]
pytest is not required unless you are actually testing things - this is fixed in the rc

unclear what version you actually have

**Comment by Rufflewind** [comment]
I had 1b52b1218e973382ea958084c47c3f33eb0ed40c installed.  Looks like it was fixed in 70ae817a5d0bd5e11e30a2e0781afa98103d656f .  Re-reinstalled Pandas and it seems to import without pytest just fine now.  Thanks!

**Pull request merged**